vpr/src/place/directed_moves_util.cpp

```diff
@@ -0,0 +1,85 @@
+#include "directed_moves_util.h"
+
+void calculate_centroid_loc(ClusterBlockId b_from, bool timing_weights, t_pl_loc& centroid, const PlacerCriticalities* criticalities) {
+    auto& device_ctx = g_vpr_ctx.device();
+    auto& grid = device_ctx.grid;
+    auto& place_ctx = g_vpr_ctx.placement();
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+
+    int x, y, pnum, ipin;
+    float acc_weight = 0;
+    float acc_x = 0;
+    float acc_y = 0;
+    float weight = 1;
+
+    //iterate over the from block pins
+    for (ClusterPinId pin_id : cluster_ctx.clb_nlist.block_pins(b_from)) {
+        ClusterNetId net_id = cluster_ctx.clb_nlist.pin_net(pin_id);
+        if (cluster_ctx.clb_nlist.net_sinks(net_id).size() == 1) {
+            ClusterBlockId source = cluster_ctx.clb_nlist.net_driver_block(net_id);
+            ClusterPinId sink_pin = *cluster_ctx.clb_nlist.net_sinks(net_id).begin();
+            ClusterBlockId sink = cluster_ctx.clb_nlist.pin_block(sink_pin);
+            if (sink == source) {
+                continue;
+            }
+        }
+
+        //if the pin is driver iterate over all the sinks
+        if (cluster_ctx.clb_nlist.pin_type(pin_id) == PinType::DRIVER) {
+            if (cluster_ctx.clb_nlist.net_is_ignored(net_id))
+                continue;
+
+            for (auto sink_pin_id : cluster_ctx.clb_nlist.net_sinks(net_id)) {
+                if (pin_id == sink_pin_id)
+                    continue;
+                ipin = cluster_ctx.clb_nlist.pin_net_index(sink_pin_id);
+                pnum = tile_pin_index(sink_pin_id);
+                if (timing_weights) {
+                    weight = criticalities->criticality(net_id, ipin);
+                } else {
+                    weight = 1;
+                }
+
+                ClusterBlockId sink_block = cluster_ctx.clb_nlist.pin_block(sink_pin_id);
+
+                x = place_ctx.block_locs[sink_block].loc.x + physical_tile_type(sink_block)->pin_width_offset[pnum];
+                y = place_ctx.block_locs[sink_block].loc.y + physical_tile_type(sink_block)->pin_height_offset[pnum];
+
+                x = std::max(std::min(x, (int)grid.width() - 2), 1);  //-2 for no perim channels
```

Comment:
As we discussed, this condition was originally copied from get_bb_from_scratch function in place.cpp. This condition was associated with the following comment:

        /* Code below counts IO blocks as being within the 1..grid.width()-2, 1..grid.height()-2 clb array. *
         * This is because channels do not go out of the 0..grid.width()-2, 0..grid.height()-2 range, and   *
         * I always take all channels impinging on the bounding box to be within   *
         * that bounding box.  Hence, this "movement" of IO blocks does not affect *
         * the which channels are included within the bounding box, and it         *
         * simplifies the code a lot.                                              */

Reply:
opened an issue to track this in a speperate context #1599